server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1678,26 +1685,58 @@ void writeTo(StreamOutput out) throws IOException {
-        private final BytesStreamOutput shardDeleteResults = new ReleasableBytesStreamOutput(bigArrays);
+        private final BytesStreamOutput shardDeleteResults;
 
         private int resultCount = 0;
+        private int skippedResultsCount = 0;
 
-        private final StreamOutput compressed = new OutputStreamStreamOutput(
-            new BufferedOutputStream(
-                new DeflaterOutputStream(Streams.flushOnCloseStream(shardDeleteResults)),
-                DeflateCompressor.BUFFER_SIZE
-            )
-        );
+        private final StreamOutput compressed;
 
         private final ArrayList<Closeable> resources = new ArrayList<>();
 
         private final ShardGenerations.Builder shardGenerationsBuilder = ShardGenerations.builder();
 
-        ShardBlobsToDelete() {
-            resources.add(compressed);
-            resources.add(LeakTracker.wrap((Releasable) shardDeleteResults));
+        public final Setting<ByteSizeValue> MAX_SHARD_DELETE_RESULTS_SIZE_SETTING = Setting.memorySizeSetting(
+            "repositories.blobstore.max_shard_delete_results_size",
+            "25%",
+            Setting.Property.NodeScope
+        );
+
+        ShardBlobsToDelete(Settings settings) {
+            this.shardDeleteResultsMaxSize = calculateMaximumShardDeleteResultsSize(settings);
+            if (this.shardDeleteResultsMaxSize > 0) {
+                this.shardDeleteResults = new ReleasableBytesStreamOutput(bigArrays);
+                this.compressed = new OutputStreamStreamOutput(
+                    new TruncatedOutputStream(
+                        new BufferedOutputStream(
+                            new DeflaterOutputStream(Streams.flushOnCloseStream(shardDeleteResults)),
+                            DeflateCompressor.BUFFER_SIZE
+                        ),
+                        shardDeleteResults::size,
+                        this.shardDeleteResultsMaxSize
+                    )
+                );
+                resources.add(compressed);
+                resources.add(LeakTracker.wrap((Releasable) shardDeleteResults));
+            } else {
+                this.shardDeleteResults = null;
+                this.compressed = null;
+            }
+        }
+
+        /**
+         * Calculates the maximum size of the shardDeleteResults BytesStreamOutput.
+         * The size should at most be 2GB, but no more than 25% of the total remaining heap space.
+         * A buffer of 1MB is maintained, so that even if the stream is of max size, there is room to flush
+         * @return The maximum number of bytes the shardDeleteResults BytesStreamOutput can consume in the heap
+         */
+        int calculateMaximumShardDeleteResultsSize(Settings settings) {
+            long maxSizeInBytes = MAX_SHARD_DELETE_RESULTS_SIZE_SETTING.get(settings).getBytes();
+            int oneMBBuffer = 1024 * 1024;
+            if (maxSizeInBytes > Integer.MAX_VALUE) {
+                return Integer.MAX_VALUE - oneMBBuffer;
```

Comment:
If `maxSizeInBytes == Integer.MAX_VALUE` then we'll skip this and return `Integer.MAX_VALUE` which doesn't allow enough headroom.